packages/script/src/script-factory.ts

```diff
@@ -0,0 +1,145 @@
+    }
+
+    this.script = new Script(
+      bytecode,
+      (args: InputValue<void>[]) =>
+        this.interface.encodeFunctionData(FUNCTION_FRAGMENT_NAME, args, 0, true),
+      (scriptResult): Result<TOutput> => {
+        const logs = getDecodedLogs(scriptResult.receipts, this.interface);
+
+        if (scriptResult.returnReceipt.type === ReceiptType.Revert) {
+          logger.throwError('Script Reverted', Logger.errors.CALL_EXCEPTION, logs);
+        }
+
+        if (
+          scriptResult.returnReceipt.type !== ReceiptType.Return &&
+          scriptResult.returnReceipt.type !== ReceiptType.ReturnData
+        ) {
+          logger.throwError(
+            `Script Return Type [${scriptResult.returnReceipt.type}] Invalid`,
+            Logger.errors.CALL_EXCEPTION,
+            {
+              logs,
+              receipt: scriptResult.returnReceipt,
+            }
+          );
+        }
+
+        let value;
+        if (scriptResult.returnReceipt.type === ReceiptType.Return) {
+          value = scriptResult.returnReceipt.val;
+        }
+        if (scriptResult.returnReceipt.type === ReceiptType.ReturnData) {
+          const decoded = this.interface.decodeFunctionResult(
+            FUNCTION_FRAGMENT_NAME,
+            scriptResult.returnReceipt.data
+          );
+          value = (decoded as [TOutput])[0];
+        }
+
+        return {
+          value,
+          logs,
+        };
+      }
+    );
+  }
+
+  async buildScriptTransaction<TData>(
+    data: InputValue<TData>[],
+    scriptOptions?: BuildScriptOptions
+  ): Promise<ScriptTransactionRequest> {
+    const options = {
+      fundTransaction: true,
+      ...scriptOptions,
+    };
+    const request = new ScriptTransactionRequest({
+      gasLimit: MAX_GAS_PER_TX,
+      ...options,
+    });
+    request.setScript(this.script, data as InputValue[]);
+
+    const requiredCoinQuantities: CoinQuantityLike[] = [];
+    if (options.fundTransaction) {
+      requiredCoinQuantities.push(request.calculateFee());
+    }
+
+    if (requiredCoinQuantities.length && this.wallet) {
+      const resources = await this.wallet.getResourcesToSpend(requiredCoinQuantities);
+      request.addResources(resources);
+    }
+
+    return request;
+  }
+
+  async call(
+    data: InputValue<TInput>[],
+    options?: BuildScriptOptions
+  ): Promise<
+    {
+      transactionResult: TransactionResult<'success' | 'failure'>;
+      response: TransactionResponse;
+    } & Result<TOutput>
+  > {
+    if (!this.provider) {
+      throw new Error('Provider is required');
+    }
+
+    const request = await this.buildScriptTransaction<TInput>(data, options);
+    const response = await this.provider.sendTransaction(request);
+    const transactionResult = await response.waitForResult();
+    const result = this.script.decodeCallResult(transactionResult);
+
+    return {
+      transactionResult,
+      response,
+      value: result.value,
+      logs: result.logs,
+    };
```

Comment:
@camsjams Quick questions:

1. Do the `Contract` calls also expose internal req/res stuff?
2. Why are they helpful in the `Script` context?

Besides, I usually like wrapping non-essential output data in a `meta` property to indicate that they are secondary.

See what you think.

```suggestion
    const request = await this.buildScriptTransaction<TInput>(data, options);
    const response = await this.provider.sendTransaction(request);
    const result = await response.waitForResult();
    const meta = { request, response, result }
    
    const { value, logs } = this.script.decodeCallResult(transactionResult);

    return { value, logs, meta }
```